Sources/NIO/PendingWritesManager.swift

```diff
@@ -447,13 +447,92 @@ internal protocol PendingWritesManager {
     var isFlushPending: Bool { get }
     var writeSpinCount: UInt { get }
     var currentBestWriteMechanism: WriteMechanism { get }
-    var channelWritabilityFlag: Atomic<Bool> { get }
+    var unwritableFlags: Atomic<UInt64> { get }
 }
 
 extension PendingWritesManager {
     // This is called from `Channel` API so must be thread-safe.
     var isWritable: Bool {
-        return self.channelWritabilityFlag.load()
+        return self.unwritableFlags.load() == 0
+    }
+
+    internal func setWritable() -> Bool {
+        // Check if flag already cleared
+        if unwritableFlags.load() & 1 == 0 {
+            return false
+        }
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue & UInt64(bitPattern: -2);
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
+        return true
+    }
+
+    internal func setUnwritable() -> Bool {
+        // Check if flag already cleared
+        if unwritableFlags.load() & 1 == 1 {
+            return false
+        }
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue | 1
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
+        return true
+    }
+
+    internal func getUserDefinedWritability(index: Int) -> Bool {
+        let mask = writabilityMask(index)
+        return (unwritableFlags.load() & mask) == mask
+    }
+
+    internal func setUserDefinedWritability(index: Int, writable: Bool) -> Bool {
+        if (writable) {
+            return setUserDefinedWritability(index: index)
+        }
+        else {
+            return clearUserDefinedWritability(index: index)
+        }
+    }
+
+    internal func setUserDefinedWritability(index: Int) -> Bool {
+        let mask = ~writabilityMask(index)
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue & mask
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
+
+        return oldValue != 0 && newValue == 0
+    }
+
+    internal func clearUserDefinedWritability(index: Int) -> Bool {
+        let mask = writabilityMask(index)
+        var oldValue: UInt64
+        var newValue: UInt64
+        repeat {
+            oldValue = unwritableFlags.load()
+            newValue = oldValue | mask
+        }
+        while !unwritableFlags.compareAndExchange(expected: oldValue, desired: newValue)
```

Comment:
This loop isn't actually necessary, and neither is the CaS. (To be fair, the CaS wasn't necessary in the original implementation either, so you were misled by what was there.)

While the writability flag is an atomic, the reason it is so is because the writability flag can be *read* from outside the event loop. It is never set from outside the event loop. Because it's only set from on the event loop, all write operations on this atomic are synchronized. That means it's just fine to replace the CaS with load/store, which has the advantage of performing vastly better, and also eliding these loops and simplifying the code.